# CommunityPowerAnalysis

Measure the functional importance of microbes residing in a microbiome community

## Installation Guide for `CommunityPowerAnalysis`  

To install and use the CommunityPowerAnalysis package in R, follow the steps below:

### Step 1: Download the Package File  
   First, download the package file CommunityPowerAnalysis_0.0.0.9000.tar.gz.   
   You do not need to clone the entire repository—just ensure you have this specific file.

   Click the file 'CommunityPowerAnalysis_0.0.0.9000.tar.gz', then click the 'Download raw file' button.  

### Step 2: Install and load the Package in R  
  Once you have the .tar.gz file, open R Studio and use the following command to install the package. Replace "path/to/CommunityPowerAnalysis_0.0.0.9000.tar.gz" with the actual path where you saved the file.  
  ```R
  install.packages("path/to/CommunityPowerAnalysis_0.0.0.9000.tar.gz", repos = NULL, type = "source")

  library(CommunityPowerAnalysis)
  ```
  For example, if the file is in your `Downloads` folder, the command would look like this:  
  ```R
  install.packages("~/Downloads/CommunityPowerAnalysis_0.0.0.9000.tar.gz", repos = NULL, type = "source")

  library(CommunityPowerAnalysis)
  ```
## Tutorial for `CommunityPowerAnalysis`
First, download and unzip the `CPA_materials` from the git  

**`CPA_materials` Directory Structure**  
This is the structure of the CPA materials directory, which contains hmmsearch result files of EARB (Extensively Acquired antimicrobial Resistant Bacteria) possessed communities.  

CPA_materials/  
├── ERR1995214/  
│   ├── ERR1995214_bin.1.fa.faa.annotation.table.txt  
│   ├── ERR1995214_bin.10.fa.faa.annotation.table.txt  
│   ├── ERR1995214_bin.2.fa.faa.annotation.table.txt  
│   ├── ERR1995214_bin.3.fa.faa.annotation.table.txt  


**Example one of the hmmsearch results**
| Gene                                    | KO     | Evalue  |
|-----------------------------------------|--------|---------|
| k115_39666_length_2143_cov_6.0000_2     | K00001 | 1.5e-74 |
| k115_30334_length_11806_cov_7.0000_6    | K00001 | 1.8e-70 |
| k115_31308_length_46843_cov_8.0000_12   | K00001 | 1.2e-61 |
| k115_16760_length_6153_cov_7.0000_1     | K00001 | 8.7e-60 |
| k115_8151_length_47996_cov_8.0000_12    | K00001 | 2.4e-48 |

### Step 1: Load Required Libraries    
Begin by loading all the necessary libraries:  
``` R
library(dplyr)
library(ggplot2)
library(data.table)
library(ggpubr)
library(CommunityPowerAnalysis)
```

### Step 2: Process KEGG Annotation Results
This function processes KEGG (Kyoto Encyclopedia of Genes and Genomes) annotation results located in multiple subdirectories. It reads files, filters out lines starting with '#' (generated by hmmsearch), and computes the count of each KEGG Orthology (KO).  

```R
# Process the KEGG results  
result_dt <- KeggResultProcess(inputDir = '/path/to/the/CPA_materials', filePattern = '.fa.faa.annotation.table.txt')
```
**Example Result Table (`result_dt`)**

| KO     | ERR1995214_bin.1 | ERR1995214_bin.10 | ERR1995214_bin.2 | ERR1995214_bin.3 | ERR1995214_bin.4 |
|--------|------------------|-------------------|------------------|------------------|------------------|
| K00001 | 14               | 3                 | 6                | 14               | 25               |
| K00002 | 9                | 1                 | 15               | 4                | 7                |
| K00003 | 3                | 2                 | 6                | 3                | 3                |
| K00004 | 9                | 2                 | 2                | 10               | 20               |  

```R
# Visualization
ggboxplot(communityPower, x="Community",y="UniqueKeggCount", fill = "Community", legend = 'none', outlier.size=0.6, size=0.1, ylab = 'Number of uniqkegg') +
  rotate_x_text(angle=45) +
  ggtitle("Number of uniqkegg") +
  rremove('x.text') +
  rremove('xlab') +
  font("ylab", size = 6)+
  font("y.text", size=6) +
  font("legend.title", size = 6)+
  font("legend.text",size = 6) +
  font("title", size = 6, face = "bold") +
  theme(legend.key.size = unit(0.3, 'cm')) +
  guides(color = guide_legend(title.position = "top",
                              title.hjust = 0.5))
```

### Step 3: Perform Unique KEGG Analysis (optional)  
The `UniqueKeggCPA` function calculates the number of unique KEGG orthologs (KOs) that are present in each bin within a community.  
This is particularly useful for assessing the functional uniqueness of different bins in a metagenomic dataset.  

```R
# Perform unique KEGG analysis
communityPower <- UniqueKeggCPA(result_dt)
```
**Example Result Table (`communityPower`)**

| Community  | Sample             | UniqueKeggCount |
|------------|--------------------|-----------------|
| ERR1995214 | ERR1995214_bin.1    | 44              |
| ERR1995214 | ERR1995214_bin.10   | 3               |
| ERR1995214 | ERR1995214_bin.2    | 14              |
| ERR1995214 | ERR1995214_bin.3    | 46              |
| ERR1995214 | ERR1995214_bin.4    | 150             |

### Step 4: Conduct Community Power Analysis  
The `CommunityPowerAnalysis` function calculates the proportion of each KEGG Ortholog (KO) that is present in each bin within a community. This is useful for understanding how specific KOs are distributed across different bins, providing insights into the functional contribution of each bin within the community.

```R
# Calculate community power scores
CPAResult <- CommunityPowerAnalysis(result_dt)

# Extract unique community and sample combinations with their power scores
CPA_scores <- CPAResult %>% select(Community, Sample, CP_Score) %>% unique()
```

**Example Result Table (`CPAResult`)**

| Community  | Sample             | KO     | CP_value | CP_Score |
|------------|--------------------|--------|----------|----------|
| ERR1995214 | ERR1995214_bin.1    | K00001 | 0.1359223| 359.3147 |
| ERR1995214 | ERR1995214_bin.1    | K00002 | 0.2142857| 359.3147 |
| ERR1995214 | ERR1995214_bin.1    | K00003 | 0.1153846| 359.3147 |
| ERR1995214 | ERR1995214_bin.1    | K00004 | 0.1323529| 359.3147 |
| ERR1995214 | ERR1995214_bin.1    | K00005 | 0.1212121| 359.3147 |  

**Example Result Table (`CPA_scores`)**

| Community  | Sample             | CP_Score |
|------------|--------------------|----------|
| ERR1995214 | ERR1995214_bin.1    | 359.3147 |
| ERR1995214 | ERR1995214_bin.10   | 171.2030 |
| ERR1995214 | ERR1995214_bin.2    | 304.7482 |
| ERR1995214 | ERR1995214_bin.3    | 467.9337 |
| ERR1995214 | ERR1995214_bin.4    | 744.2165 |

```R
# Visualization
ggboxplot(CPA_scores, x="Community",y="CP_Score", fill = "Community", legend = 'none', outlier.size=0.6, size=0.1, ylab = 'Community power score') +
  rotate_x_text(angle=45) +
  ggtitle("Community power analysis") +
  rremove('x.text') +
  rremove('xlab') +
  # font("xlab", size = 6)+
  font("ylab", size = 6)+
  font("y.text", size=6) +
  # font("xy.text", size = 6)+
  font("legend.title", size = 6)+
  font("legend.text",size = 6) +
  font("title", size = 6, face = "bold") +
  theme(legend.key.size = unit(0.3, 'cm')) +
  guides(color = guide_legend(title.position = "top",
                              title.hjust = 0.5))
```
